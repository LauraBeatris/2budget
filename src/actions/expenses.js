import database from "../firebase/firebase";

//ADD_EXPENSE
export const addExpense = expense => {
  return {
    type: "ADD_EXPENSE",
    expense
  };
};

export const addExpenseRequest = (expenseData = {}) => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;

    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = {
      description,
      note,
      amount,
      createdAt
    };
    // After sending this to database -> dispatch the action to the store
    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

//EDIT_EXPENSE
export const editExpense = (id, update) => ({
  type: "EDIT_EXPENSE",
  id,
  update
});

//SET_EXPENSE
export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const expenses = [];
    const { uid } = getState().auth;

    return database
      .ref(`users/${uid}/expenses`)
      .once("value")
      .then(snapshot => {
        snapshot.forEach(childSnapshot => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });

        dispatch(setExpenses(expenses));
      });
  };
};

export const startRemoveExpenses = ({ id } = {}) => {
  // This asyncronous function is responsable for doing a connection with
  // firebase and dispatching an action
  return (dispatch, getState) => {
    const { uid } = getState().auth;

    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(snapshot => {
        //console.log("expense removed");
        dispatch(removeExpense({ id }));
      });
  };
};

export const startEditExpenses = (id, update) => {
  return (dispatch, getState) => {
    const { uid } = getState().auth;

    // Getting the reference of the expense from database and setting the new value
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update(update)
      .then(snapshot => {
        //expense edit

        //Dispatching the action to store
        dispatch(editExpense(id, update));
      })
      .catch(err => console.log(err));
  };
};
