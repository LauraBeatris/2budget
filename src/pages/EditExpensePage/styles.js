import styled from "styled-components";

export const ButtonRemove = styled.button`
  margin-top: 0.5rem;
  background-color: #6dbfcc;
  border: none;
  font-size: 1.6rem;
  font-weight: bold;
  padding: 2rem 4rem;
  color: white;
  border-radius: 20px;
  box-shadow: 0 1rem 2rem rgba(0, 0, 0, 0.2);
  cursor: pointer;
  text-decoration: none;
  transition: all 0.5s;

  &:hover {
    transform: translateY(-6px);
  }

  &:active {
    transform: translateY(0px);
    box-shadow: 0 2rem 2rem rgba(0, 0, 0, 0.2);
  }
`;

export const EditContainer = styled.div`
  text-align: center;
  margin-top: 5rem;
`;

export const ModalContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  text-align: center;
  padding: 4rem;

  p {
    color: #6dbfcc;
    font-weight: 600;
    font-size: 1.7rem;
  }

  div {
    margin-top: 1.5rem;
    display: inherit;
    flex-direction: row;
    width: 50%;
    justify-content: space-between;

    button {
      background-color: white;
      box-shadow: 0 1rem 2rem rgba(0, 0, 0, 0.1);
      border-radius: 10px;
      border: 1px solid rgb(77, 81, 94);
      color: #6dbfcc;
      padding: 1rem 3rem;
      font-weight: 500;
      transition: all 0.35s;

      &:hover {
        cursor: pointer;
        transform: translateY(-6px);
      }
    }
  }
`;
