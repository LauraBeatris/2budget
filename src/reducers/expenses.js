const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.update // overriden the expense items with the action updates icons
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};

export default expensesReducer;
