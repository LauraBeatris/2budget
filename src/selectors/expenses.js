import moment from "moment";

// Get visible expenses
export const getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return b.createdAt < a.createdAt ? -1 : 1;
      } else if (sortBy === "amount") {
        return a.amount > b.amount ? -1 : 1;
      } else {
        return null;
      }
    });
};

export const getExpenseTotal = expenses => {
  if (expenses.length === 0) {
    return 0;
  } else if (expenses.length > 1) {
    const amounts = expenses.map(expense => expense.amount);
    return amounts.reduce((sum, value) => sum + value, 1);
  } else {
    return expenses[0].amount;
  }
};
