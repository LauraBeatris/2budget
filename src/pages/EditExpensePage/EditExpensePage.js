import React, { Component } from "react";
import Helmet from "react-helmet";
import Header from "../../components/Header/Header";
import { connect } from "react-redux";
import ExpenseForm from "../../components/ExpenseForm/ExpenseForm";
import { startRemoveExpenses, startEditExpenses } from "../../actions/expenses";
import { ButtonRemove, EditContainer } from "./styles";

export class EditExpensePage extends Component {
  onSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/dashboard");
  };

  onClick = () => {
    this.props.removeExpense(this.props.expense);
    this.props.history.push("/dashboard");
  };

  render() {
    return (
      <div>
        <Header />
        <Helmet>
          <title> 2Budget | Edit Expense </title>
        </Helmet>
        <EditContainer>
          <ExpenseForm
            type="Edit"
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <ButtonRemove id="remove-button" onClick={this.onClick}>
            Remove
          </ButtonRemove>
        </EditContainer>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    // Finding the expense with the match id
    expense: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

const mapDispatchToProps = dispatch => ({
  editExpense: (id, expense) => dispatch(startEditExpenses(id, expense)),
  removeExpense: expense => dispatch(startRemoveExpenses(expense))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
