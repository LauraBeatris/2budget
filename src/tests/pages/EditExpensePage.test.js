import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../pages/EditExpensePage/EditExpensePage";
import expenses from "../fixtures/expenses";

let wrapper, expense, history, editExpense, startRemoveExpenses;

beforeEach(() => {
  history = {
    push: jest.fn()
  };
  expense = expenses[0];
  startRemoveExpenses = jest.fn();
  editExpense = jest.fn();
  wrapper = shallow(
    <EditExpensePage
      history={history}
      expense={expense}
      removeExpense={startRemoveExpenses}
      editExpense={editExpense}
    />
  );
});

test("should render EditExpansePage correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle editExpense", () => {
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[0]);
  expect(editExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith("/dashboard");
});

test("should handle removeExpense", () => {
  wrapper.find("#remove-button").simulate("click");
  expect(startRemoveExpenses).toHaveBeenLastCalledWith(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith("/dashboard");
});
