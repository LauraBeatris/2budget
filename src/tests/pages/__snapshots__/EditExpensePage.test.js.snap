// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpansePage correctly 1`] = `
<div>
  <HelmetWrapper
    defer={true}
    encodeSpecialCharacters={true}
  >
    <title>
       2Budget | Edit Expense 
    </title>
  </HelmetWrapper>
  <.sc-htpNat>
    <ExpenseForm
      expense={
        Object {
          "amount": 195,
          "createdAt": "1970-01-01T00:00:00.000Z",
          "description": "Water rent",
          "id": "1",
          "note": "Random note",
        }
      }
      onSubmit={[Function]}
      type="Edit"
    />
    <.sc-bwzfXH
      id="remove-button"
      onClick={[Function]}
    >
      Remove
    </.sc-bwzfXH>
  </.sc-htpNat>
</div>
`;
